### PR TITLE
Skip load `config.yml` config

### PR DESCRIPTION
**Pull request by LoicMahieu**
**Summary**

I would like to make config more "programmable". Collections may comes from functions, uses reusable fields etc...

With this PR, we'll be able to specify config only by JS with a manual init.

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 9b2ae0f5c3c85f47f6ceee4b1c449303c4eeb07d

https://deploy-preview-1866--netlify-cms-www.netlify.com

**Comment by talves** [comment]
Can you explain this change and what it changes? Not sure I am following how your change improves the current process.

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit bc057013c169823d6752489fa4b7689b7a98ca26

https://deploy-preview-1866--cms-demo.netlify.com

File: packages/netlify-cms-core/src/actions/config.js [hunk 1]
```diff
@@ -111,12 +106,15 @@ export function loadConfig() {
     try {
       const preloadedConfig = getState().config;
       const configUrl = getConfigUrl();
-      const loadedConfig = await getConfig(configUrl, preloadedConfig && preloadedConfig.size > 1);
+      const loadedConfig =
+        bootstrapConfig && bootstrapConfig.loadConfig === false
```
Comment (talves) [inline packages/netlify-cms-core/src/actions/config.js]: The conditional here will default to not use the `preloadedConfig` and break my current projects.

Good idea to not try loading the `config.yml` upon conditional config though. I think it should be something like

    bootstrapConfig && bootstrapConfig.ignorePreloadConfig
          ? {}
          : await getConfig(configUrl, preloadedConfig && preloadedConfig.size > 1);

cc/ @erquhart

Comment (talves) [reply packages/netlify-cms-core/src/actions/config.js]: I think that will work.

File: packages/netlify-cms-core/src/actions/config.js [hunk 2]
```diff
@@ -40,9 +40,8 @@ export function applyDefaults(config) {
     });
 }
 
-function mergePreloadedConfig(preloadedConfig, loadedConfig) {
-  const map = fromJS(loadedConfig) || Map();
-  return preloadedConfig ? preloadedConfig.mergeDeep(map) : map;
+function mergePreloadedConfig(...loadedConfig) {
+  return Map().merge(...loadedConfig);
```
Comment (erquhart) [inline packages/netlify-cms-core/src/actions/config.js]: We should continue using `mergeDeep` for this.

File: packages/netlify-cms-core/src/actions/config.js [hunk 3]
```diff
@@ -111,12 +106,15 @@ export function loadConfig() {
     try {
       const preloadedConfig = getState().config;
       const configUrl = getConfigUrl();
-      const loadedConfig = await getConfig(configUrl, preloadedConfig && preloadedConfig.size > 1);
+      const loadedConfig =
+        bootstrapConfig && bootstrapConfig.ignoreLoadConfig
```
Comment (erquhart) [inline packages/netlify-cms-core/src/actions/config.js]: Snake case is used for our yaml config keys, and we try to phrase keys in the positive, so this should probably be `load_config_file`, defaulting to `true`, and we check for a `false` value to skip loading.

File: packages/netlify-cms-core/src/actions/config.js [hunk 4]
```diff
@@ -111,12 +106,15 @@ export function loadConfig() {
     try {
       const preloadedConfig = getState().config;
```
Comment (erquhart) [inline packages/netlify-cms-core/src/actions/config.js]: I don't believe this has a role any longer with the changes in this PR.

Comment (LoicMahieu) [reply packages/netlify-cms-core/src/actions/config.js]: Surely, I confess that I can't say if it is still relevant. Do you confirm that we can remove this ?

Comment (erquhart) [reply packages/netlify-cms-core/src/actions/config.js]: Yeah, passing `bootstrapConfig` into `loadConfig` supersedes the old `preloadedConfig` behavior, you can drop it.

**Comment by erquhart** [comment]
@talves would you have time to give this a quick review?

**Comment by talves** [comment]
@erquhart isn't this still waiting for your requested changes?

**Comment by erquhart** [comment]
He added commits to resolve those.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by talves** [comment]
https://github.com/netlify/netlify-cms/pull/1866/commits/bea25af9121180b45dbe5d08548858848471a587#r235578398

That has been resolved?

**Comment by erquhart** [comment]
Ah, that one is just eliminating dead code, no actual impact on behavior.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by LoicMahieu** [comment]
> I just can't get my head around why we want to pass the bootstrapConfig object to the App component as a prop to make this change. We can already load a config using the import of init/bootstrap.

Indeed but... I seems really ugly.

> Really good point @talves, didn't even notice that. @LoicMahieu any thoughts on Tony's comment? We should remove any unnecessary changes to reduce risk.

I confess that I am lacking of time to refactor this changes in a new branch with only these "small" changes.
Furthermore, I don't think theses changes introduce a breaking change (but certainly should be released as minor). I agree that we should minimize the risk to break something but something, it is necessary to go ahead.

**Comment by talves** [comment]
I agree. 

Do you want me to send a PR to only do what I said [in this comment](https://github.com/netlify/netlify-cms/pull/1866#issuecomment-441390153), since it would not be a breaking change?

**Comment by erquhart** [comment]
Closing in favor of #2053.

**Pull request closed without merging**